File: samples/javalin-sureness/src/main/java/org/dromara/sureness/javalin/Application.java
package org.dromara.sureness.javalin;

import io.javalin.Javalin;

import org.dromara.sureness.DefaultSurenessConfig;
import org.dromara.sureness.mgt.SurenessSecurityManager;
import org.dromara.sureness.processor.exception.*;
import org.dromara.sureness.subject.SubjectSum;
import org.dromara.sureness.util.JsonWebTokenUtil;
import org.dromara.sureness.util.SurenessContextHolder;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.List;
import java.util.UUID;

import static io.javalin.apibuilder.ApiBuilder.*;

/**
 * @author tomsun28
 * @date 2020-09-15 23:59
 */
public class Application {

    private static final Logger log = LoggerFactory.getLogger(Application.class);

    @SuppressWarnings("unchecked")
    public static void main(String[] args) {
        // init sureness default config
        new DefaultSurenessConfig();

        // init javalin
        Javalin app = Javalin.create().start(8088);

        // intercept all rest requests for authenticating and authorizing
        app.before(ctx -> {
            SubjectSum subject = SurenessSecurityManager.getInstance().checkIn(ctx.req);
            // when auth error , the exception throw, you should use app.exception() catch it and define return
            if (subject != null) {
                SurenessContextHolder.bindSubject(subject);
            }
        });

        app.after(ctx ->  SurenessContextHolder.unbindSubject());

        // when auth error , the exception throw, you should use app.exception() catch it and define return
        app.exception(UnknownAccountException.class, (e, ctx) -> {
            log.debug("this request user account not exist");
            ctx.status(401).result(e.getMessage());
        }).exception(IncorrectCredentialsException.class, (e, ctx) -> {
            log.debug("this account credential is incorrect");
            ctx.status(401).result(e.getMessage());
        }).exception(ExpiredCredentialsException.class, (e, ctx) -> {
            log.debug("this account credential expired");
            ctx.status(401).result(e.getMessage());
        }).exception(NeedDigestInfoException.class, (e, ctx) -> {
            log.debug("you should try once again with digest auth information");
            ctx.status(401).header("WWW-Authenticate", e.getAuthenticate());
        }).exception(UnauthorizedException.class, (e, ctx) -> {
            log.debug("this account can not access this resource");
            ctx.status(403).result(e.getMessage());
        }).exception(Exception.class, (e, ctx) -> {
            log.error("other exception happen: ", e);
            ctx.status(500).result(e.getMessage());
        });

        // issue jwt rest api
        app.get("/auth/token", ctx -> {
            SubjectSum subjectSum = SurenessContextHolder.getBindSubject();
            if (subjectSum == null) {
                ctx.result("Please auth!");
            } else {
                String principal = (String) subjectSum.getPrincipal();
                List<String> roles = (List<String>) subjectSum.getRoles();
                // issue jwt
                String jwt = JsonWebTokenUtil.issueJwt(UUID.randomUUID().toString(), principal,
                        "token-server", 3600L, roles);
                ctx.result(jwt);
            }
        });

        // simple rest api
        app.routes(() ->
            path("api", () -> {
                path("v3", () -> {
                    get("host", ctx -> ctx.result("get /api/v3/host success"));
                    put("book", ctx -> ctx.result("put /api/v3/book success"));
                });
                path("v2", () -> {
                    path("host", () -> {
                        get(ctx -> ctx.result("get /api/v2/host success"));
                        post(ctx -> ctx.result("post /api/v2/host success"));
                        put(ctx -> ctx.result("put /api/v2/host success"));
                        delete(ctx -> ctx.result("delete /api/v2/host success"));
                    });
                });
                path("v1", () -> {
                    path("source1", () -> {
                        get(ctx -> ctx.result("get /api/v1/source1 success"));
                        post(ctx -> ctx.result("post /api/v1/source1 success"));
                        put(ctx -> ctx.result("put /api/v1/source1 success"));
                        delete(ctx -> ctx.result("delete /api/v1/source1 success"));
                    });
                });
            }));
    }
}

File: samples/ktor-netty-sureness/src/Application.kt
package com.usthe.sureness.demo

import com.usthe.sureness.DefaultSurenessConfig
import com.usthe.sureness.mgt.SurenessSecurityManager
import com.usthe.sureness.processor.exception.*
import com.usthe.sureness.subject.support.PasswordSubject
import com.usthe.sureness.util.SurenessContextHolder
import io.ktor.application.Application
import io.ktor.application.ApplicationCallPipeline
import io.ktor.application.call
import io.ktor.application.log
import io.ktor.http.ContentType
import io.ktor.response.respondText
import io.ktor.routing.*
import io.ktor.server.engine.EngineAPI
import io.ktor.server.engine.embeddedServer
import io.ktor.server.netty.Netty



@EngineAPI
fun main(args: Array<String>) {
    embeddedServer(Netty, port = 8081){main()}.start(wait = true)
}


@EngineAPI
fun Application.main() {
    // init sureness default config
    DefaultSurenessConfig()

    intercept(ApplicationCallPipeline.Call) {
        try {
            val request = call.request
            val subject = SurenessSecurityManager.getInstance().checkIn(request)
            // if auth success contextHolder bind the subject
            if (subject != null) {
                SurenessContextHolder.bindSubject(subject)
                log.debug("auth success!")
            }
        } catch (e4: ProcessorNotFoundException) {
            log.debug("this request is illegal")
            call.respondText(e4.localizedMessage)
            return@intercept finish()

        } catch (e4: UnknownAccountException) {
            log.debug("this request is illegal")
            call.respondText(e4.localizedMessage)
            return@intercept finish()

        } catch (e4: UnsupportedSubjectException) {
            log.debug("this request is illegal")
            call.respondText(e4.localizedMessage)
            return@intercept finish()

        } catch (e2: DisabledAccountException) {
            log.debug("the account is disabled")
            call.respondText(e2.localizedMessage)
            return@intercept finish()

        } catch (e2: ExcessiveAttemptsException) {
            log.debug("the account is disabled")
            call.respondText(e2.localizedMessage)
            return@intercept finish()

        } catch (e3: IncorrectCredentialsException) {
            log.debug("this account credential is incorrect or expired")
            call.respondText(e3.localizedMessage)
            return@intercept finish()

        } catch (e3: ExpiredCredentialsException) {
            log.debug("this account credential is incorrect or expired")
            call.respondText(e3.localizedMessage)
            return@intercept finish()

        } catch (e5: UnauthorizedException) {
            log.debug("this account can not access this resource")
            call.respondText(e5.localizedMessage)
            return@intercept finish()

        } catch (e: RuntimeException) {
            log.error("other exception happen: ", e)
            call.respondText(e.localizedMessage)
            return@intercept finish()
        }
    }

    routing {
        get("/api/v3/host") {
            call.respondText("Hello World!", ContentType.Text.Plain)
        }
        get("/api/v2/host") {
            call.respondText("get /api/v2/host")
        }
        post("/api/v2/host") {
            call.respondText("post /api/v2/host")
        }
        put("/api/v2/host") {
            call.respondText("put /api/v2/host")
        }
        delete("/api/v2/host") {
            call.respondText("delete /api/v2/host")
        }
        put("/api/mi/tom") {
            call.respondText("put /api/mi/tom")
        }
        get("/api/v1/getSource1") {
            call.respondText("get /api/v1/getSource1")
        }
        get("/api/v2/getSource2/book") {
            call.respondText("get /api/v2/getSource2/book")
        }
        get("/api/v1/source1") {
            call.respondText("get /api/v1/source1")
        }
        post("/api/v1/source1") {
            call.respondText("post /api/v1/source1")
        }
        put("/api/v1/source1") {
            call.respondText("put /api/v1/source1")
        }
        delete("/api/v1/source1") {
            call.respondText("delete /api/v1/source1")
        }
    }
}